### PR TITLE
Bulk assign permissions for multiple objects

### PR DESCRIPTION
**Pull request by phinjensen**
I just started using django-guardian and found myself really wanting a function to assign permissions for multiple objects at once. Then I found issue #52 and decided I should write it myself. Anyway, here's what I've come up with.

I'm also working on writing an action for `GuardedModelAdmin` to work with this function, but that'll come later.

### CODE REVIEW

**Comment by brianmay** [comment]
@ad-m What do you think of this proposed change?

**Comment by ad-m** [comment]
I look at it today. I have to consider using a ```bulk_create```.

**Comment by phinjensen** [comment]
If you would prefer it to be in a separate function, I'd be happy to re-write it like that.

**Comment by ad-m** [comment]
There is a lot of queries. Are they all necessary?
```
In [26]: assign_perm('posts.change_post', User.objects.last(), Post.objects.all()[:5])
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" LIMIT 5; args=()
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" LIMIT 5; args=()
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE ("django_content_type"."model" = 'post' AND "django_content_type"."app_label" = 'posts'); args=('post', 'posts')
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE ("django_content_type"."model" = 'post' AND "django_content_type"."app_label" = 'posts'); args=('post', 'posts')
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'change_post' AND "auth_permission"."content_type_id" = 9); args=(u'change_post', 9)
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'change_post' AND "auth_permission"."content_type_id" = 9); args=(u'change_post', 9)
(0.000) SELECT "guardian_userobjectpermission"."id", "guardian_userobjectpermission"."permission_id", "guardian_userobjectpermission"."content_type_id", "guardian_userobjectpermission"."object_pk", "guardian_userobjectpermission"."user_id" FROM "guardian_userobjectpermission" WHERE ("guardian_userobjectpermission"."object_pk" = '1' AND "guardian_userobjectpermission"."user_id" = 4 AND "guardian_userobjectpermission"."content_type_id" = 9 AND "guardian_userobjectpermission"."permission_id" = 26); args=(u'1', 4, 9, 26)
(0.000) SELECT "guardian_userobjectpermission"."id", "guardian_userobjectpermission"."permission_id", "guardian_userobjectpermission"."content_type_id", "guardian_userobjectpermission"."object_pk", "guardian_userobjectpermission"."user_id" FROM "guardian_userobjectpermission" WHERE ("guardian_userobjectpermission"."object_pk" = '1' AND "guardian_userobjectpermission"."user_id" = 4 AND "guardian_userobjectpermission"."content_type_id" = 9 AND "guardian_userobjectpermission"."permission_id" = 26); args=(u'1', 4, 9, 26)
(0.000) BEGIN; args=None
(0.000) BEGIN; args=None
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" WHERE "posts_post"."id" = 1; args=(1,)
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" WHERE "posts_post"."id" = 1; args=(1,)
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 9; args=(9,)
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 9; args=(9,)
(0.000) INSERT INTO "guardian_userobjectpermission" ("permission_id", "content_type_id", "object_pk", "user_id") VALUES (26, 9, '1', 4); args=[26, 9, u'1', 4]
(0.000) INSERT INTO "guardian_userobjectpermission" ("permission_id", "content_type_id", "object_pk", "user_id") VALUES (26, 9, '1', 4); args=[26, 9, u'1', 4]
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'change_post' AND "auth_permission"."content_type_id" = 9); args=(u'change_post', 9)
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'change_post' AND "auth_permission"."content_type_id" = 9); args=(u'change_post', 9)
(0.000) SELECT "guardian_userobjectpermission"."id", "guardian_userobjectpermission"."permission_id", "guardian_userobjectpermission"."content_type_id", "guardian_userobjectpermission"."object_pk", "guardian_userobjectpermission"."user_id" FROM "guardian_userobjectpermission" WHERE ("guardian_userobjectpermission"."object_pk" = '2' AND "guardian_userobjectpermission"."user_id" = 4 AND "guardian_userobjectpermission"."content_type_id" = 9 AND "guardian_userobjectpermission"."permission_id" = 26); args=(u'2', 4, 9, 26)
(0.000) SELECT "guardian_userobjectpermission"."id", "guardian_userobjectpermission"."permission_id", "guardian_userobjectpermission"."content_type_id", "guardian_userobjectpermission"."object_pk", "guardian_userobjectpermission"."user_id" FROM "guardian_userobjectpermission" WHERE ("guardian_userobjectpermission"."object_pk" = '2' AND "guardian_userobjectpermission"."user_id" = 4 AND "guardian_userobjectpermission"."content_type_id" = 9 AND "guardian_userobjectpermission"."permission_id" = 26); args=(u'2', 4, 9, 26)
(0.000) BEGIN; args=None
(0.000) BEGIN; args=None
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" WHERE "posts_post"."id" = 2; args=(2,)
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" WHERE "posts_post"."id" = 2; args=(2,)
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 9; args=(9,)
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 9; args=(9,)
(0.000) INSERT INTO "guardian_userobjectpermission" ("permission_id", "content_type_id", "object_pk", "user_id") VALUES (26, 9, '2', 4); args=[26, 9, u'2', 4]
(0.000) INSERT INTO "guardian_userobjectpermission" ("permission_id", "content_type_id", "object_pk", "user_id") VALUES (26, 9, '2', 4); args=[26, 9, u'2', 4]
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'change_post' AND "auth_permission"."content_type_id" = 9); args=(u'change_post', 9)
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'change_post' AND "auth_permission"."content_type_id" = 9); args=(u'change_post', 9)
(0.000) SELECT "guardian_userobjectpermission"."id", "guardian_userobjectpermission"."permission_id", "guardian_userobjectpermission"."content_type_id", "guardian_userobjectpermission"."object_pk", "guardian_userobjectpermission"."user_id" FROM "guardian_userobjectpermission" WHERE ("guardian_userobjectpermission"."object_pk" = '3' AND "guardian_userobjectpermission"."user_id" = 4 AND "guardian_userobjectpermission"."content_type_id" = 9 AND "guardian_userobjectpermission"."permission_id" = 26); args=(u'3', 4, 9, 26)
(0.000) SELECT "guardian_userobjectpermission"."id", "guardian_userobjectpermission"."permission_id", "guardian_userobjectpermission"."content_type_id", "guardian_userobjectpermission"."object_pk", "guardian_userobjectpermission"."user_id" FROM "guardian_userobjectpermission" WHERE ("guardian_userobjectpermission"."object_pk" = '3' AND "guardian_userobjectpermission"."user_id" = 4 AND "guardian_userobjectpermission"."content_type_id" = 9 AND "guardian_userobjectpermission"."permission_id" = 26); args=(u'3', 4, 9, 26)
(0.000) BEGIN; args=None
(0.000) BEGIN; args=None
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" WHERE "posts_post"."id" = 3; args=(3,)
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" WHERE "posts_post"."id" = 3; args=(3,)
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 9; args=(9,)
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 9; args=(9,)
(0.000) INSERT INTO "guardian_userobjectpermission" ("permission_id", "content_type_id", "object_pk", "user_id") VALUES (26, 9, '3', 4); args=[26, 9, u'3', 4]
(0.000) INSERT INTO "guardian_userobjectpermission" ("permission_id", "content_type_id", "object_pk", "user_id") VALUES (26, 9, '3', 4); args=[26, 9, u'3', 4]
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'change_post' AND "auth_permission"."content_type_id" = 9); args=(u'change_post', 9)
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'change_post' AND "auth_permission"."content_type_id" = 9); args=(u'change_post', 9)
(0.000) SELECT "guardian_userobjectpermission"."id", "guardian_userobjectpermission"."permission_id", "guardian_userobjectpermission"."content_type_id", "guardian_userobjectpermission"."object_pk", "guardian_userobjectpermission"."user_id" FROM "guardian_userobjectpermission" WHERE ("guardian_userobjectpermission"."object_pk" = '4' AND "guardian_userobjectpermission"."user_id" = 4 AND "guardian_userobjectpermission"."content_type_id" = 9 AND "guardian_userobjectpermission"."permission_id" = 26); args=(u'4', 4, 9, 26)
(0.000) SELECT "guardian_userobjectpermission"."id", "guardian_userobjectpermission"."permission_id", "guardian_userobjectpermission"."content_type_id", "guardian_userobjectpermission"."object_pk", "guardian_userobjectpermission"."user_id" FROM "guardian_userobjectpermission" WHERE ("guardian_userobjectpermission"."object_pk" = '4' AND "guardian_userobjectpermission"."user_id" = 4 AND "guardian_userobjectpermission"."content_type_id" = 9 AND "guardian_userobjectpermission"."permission_id" = 26); args=(u'4', 4, 9, 26)
(0.000) BEGIN; args=None
(0.000) BEGIN; args=None
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" WHERE "posts_post"."id" = 4; args=(4,)
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" WHERE "posts_post"."id" = 4; args=(4,)
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 9; args=(9,)
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 9; args=(9,)
(0.000) INSERT INTO "guardian_userobjectpermission" ("permission_id", "content_type_id", "object_pk", "user_id") VALUES (26, 9, '4', 4); args=[26, 9, u'4', 4]
(0.000) INSERT INTO "guardian_userobjectpermission" ("permission_id", "content_type_id", "object_pk", "user_id") VALUES (26, 9, '4', 4); args=[26, 9, u'4', 4]
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'change_post' AND "auth_permission"."content_type_id" = 9); args=(u'change_post', 9)
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'change_post' AND "auth_permission"."content_type_id" = 9); args=(u'change_post', 9)
(0.000) SELECT "guardian_userobjectpermission"."id", "guardian_userobjectpermission"."permission_id", "guardian_userobjectpermission"."content_type_id", "guardian_userobjectpermission"."object_pk", "guardian_userobjectpermission"."user_id" FROM "guardian_userobjectpermission" WHERE ("guardian_userobjectpermission"."object_pk" = '5' AND "guardian_userobjectpermission"."user_id" = 4 AND "guardian_userobjectpermission"."content_type_id" = 9 AND "guardian_userobjectpermission"."permission_id" = 26); args=(u'5', 4, 9, 26)
(0.000) SELECT "guardian_userobjectpermission"."id", "guardian_userobjectpermission"."permission_id", "guardian_userobjectpermission"."content_type_id", "guardian_userobjectpermission"."object_pk", "guardian_userobjectpermission"."user_id" FROM "guardian_userobjectpermission" WHERE ("guardian_userobjectpermission"."object_pk" = '5' AND "guardian_userobjectpermission"."user_id" = 4 AND "guardian_userobjectpermission"."content_type_id" = 9 AND "guardian_userobjectpermission"."permission_id" = 26); args=(u'5', 4, 9, 26)
(0.000) BEGIN; args=None
(0.000) BEGIN; args=None
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" WHERE "posts_post"."id" = 5; args=(5,)
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" WHERE "posts_post"."id" = 5; args=(5,)
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 9; args=(9,)
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 9; args=(9,)
(0.000) INSERT INTO "guardian_userobjectpermission" ("permission_id", "content_type_id", "object_pk", "user_id") VALUES (26, 9, '5', 4); args=[26, 9, u'5', 4]
(0.000) INSERT INTO "guardian_userobjectpermission" ("permission_id", "content_type_id", "object_pk", "user_id") VALUES (26, 9, '5', 4); args=[26, 9, u'5', 4]
```

**Comment by ad-m** [comment]
I think it is a good idea to prepare a seperate function. There is a lot of job to fix performance issue. We should prefetch Permission object for QuerySet. It can be achieved easily eg.
```
diff --git a/guardian/managers.py b/guardian/managers.py
index b441b75..f78acc2 100644
--- a/guardian/managers.py
+++ b/guardian/managers.py
@@ -31,8 +31,12 @@ class UserObjectPermissionManager(BaseObjectPermissionManager):
         if getattr(obj, 'pk', None) is None:
             raise ObjectNotPersisted("Object %s needs to be persisted first"
                                      % obj)
+
         ctype = ContentType.objects.get_for_model(obj)
-        permission = Permission.objects.get(content_type=ctype, codename=perm)
+        if not isinstance(perm, Permission):
+            permission = Permission.objects.get(content_type=ctype, codename=perm)
+        else:
+            permission = perm
 
         kwargs = {'permission': permission, 'user': user}
         if self.is_generic():
@@ -82,7 +86,11 @@ class GroupObjectPermissionManager(BaseObjectPermissionManager):
             raise ObjectNotPersisted("Object %s needs to be persisted first"
                                      % obj)
         ctype = ContentType.objects.get_for_model(obj)
-        permission = Permission.objects.get(content_type=ctype, codename=perm)
+        if not isinstance(perm, Permission):
+            ctype = ContentType.objects.get_for_model(obj)
+            permission = Permission.objects.get(content_type=ctype, codename=perm)
+        else:
+            permission = perm
 
         kwargs = {'permission': permission, 'group': group}
         if self.is_generic():
diff --git a/guardian/shortcuts.py b/guardian/shortcuts.py
index c2274b5..06c89e8 100644
--- a/guardian/shortcuts.py
+++ b/guardian/shortcuts.py
@@ -6,11 +6,11 @@ from __future__ import unicode_literals
 from django.contrib.auth.models import Group
 from django.contrib.auth.models import Permission
 from django.contrib.contenttypes.models import ContentType
-from django.db.models import Count, Q
+from django.db.models import Count, Q, QuerySet
 from django.apps import apps
 from django.shortcuts import _get_queryset
+from django.utils import six
 from itertools import groupby
-
 from guardian.compat import basestring
 from guardian.compat import get_user_model
 from guardian.core import ObjectPermissionChecker
@@ -28,8 +28,9 @@ def assign_perm(perm, user_or_group, obj=None):
     Assigns permission to user/group and object pair(s).
 
     :param perm: proper permission for given ``obj``, as string (in format:
-      ``app_label.codename`` or ``codename``). If ``obj`` is not given, must
-      be in format ``app_label.codename``.
+      ``app_label.codename`` or ``codename``) or ``Permission`` instance.
+      If ``obj`` is not given, must be in format ``app_label.codename`` or
+      ``Permission`` instance.
 
     :param user_or_group: instance of ``User``, ``AnonymousUser`` or ``Group``;
       passing any other object would raise
@@ -83,13 +84,15 @@ def assign_perm(perm, user_or_group, obj=None):
     user, group = get_identity(user_or_group)
     # If obj is None we try to operate on global permissions
     if obj is None:
-        try:
-            app_label, codename = perm.split('.', 1)
-        except ValueError:
-            raise ValueError("For global permissions, first argument must be in"
-                             " format: 'app_label.codename' (is %r)" % perm)
-        perm = Permission.objects.get(content_type__app_label=app_label,
-                                      codename=codename)
+        if not isinstance(perm, Permission):
+            try:
+                app_label, codename = perm.split('.', 1)
+            except ValueError:
+                raise ValueError("For global permissions, first argument must be in"
+                                 " format: 'app_label.codename' (is %r)" % perm)
+            perm = Permission.objects.get(content_type__app_label=app_label,
+                                          codename=codename)
+
         if user:
             user.user_permissions.add(perm)
             return perm
@@ -97,7 +100,14 @@ def assign_perm(perm, user_or_group, obj=None):
             group.permissions.add(perm)
             return perm
 
-    perm = perm.split('.')[-1]
+    if not isinstance(perm, Permission):
+        perm = perm.split('.')[-1]
+
+    if isinstance(obj, QuerySet) and not isinstance(perm, Permission):
+        ctype = ContentType.objects.get_for_model(obj.model)
+        perm = Permission.objects.get(content_type=ctype, codename=perm)
+        return assign_perm(perm, user_or_group, obj)
+
     try:
         assigned_perms = []
         for instance in obj:
```

But it might be a good idea to use a ```bulk_create``` for create ```*_obj_perms_model`` instance. If we use ```bulk_create``` we have hand some remove duplicates instead use ```get_or_create```, so we have fetch permission for objects, then skip objects if exists. 

Operation of all these aspects in one function will greatly increase the complexity of ```assign_perm```.

**Comment by phinjensen** [comment]
@ad-m I agree, the `assign_perm` and `remove_perm` functions would probably get very complex if we started optimizing like that. I moved the queryset functionality of those into separate functions with your permission optimization, which cuts down on queries a little bit.

I haven't started working on optimizing to use bulk_create to reduce queries yet, but is it necessary to optimize that right now? As it is, won't this function as well as (or better) than a manual for loop?

**Comment by ad-m** [comment]
@phinjensen , For me, the code is correct and now brings something additional, so it can be merged.

**Comment by phinjensen** [comment]
@brianmay Can you merge this?

File: guardian/shortcuts.py [hunk 1]
```diff
@@ -113,7 +148,7 @@ def remove_perm(perm, user_or_group=None, obj=None):
       passing any other object would raise
       ``guardian.exceptions.NotUserNorGroup`` exception
 
-    :param obj: persisted Django's ``Model`` instance or ``None`` if assigning
+    :param obj: persisted Django's ``Model`` instance, or ``None`` if assigning
```
Comment (ad-m) [inline guardian/shortcuts.py]: Why comma added?

**Comment by ad-m** [comment]
I optimized ```bulk_assign_perm```  and ```bulk_remove_perm```.

```diff
diff --git a/guardian/core.py b/guardian/core.py
index 53fb1c2..d9e8c4f 100644
--- a/guardian/core.py
+++ b/guardian/core.py
@@ -72,11 +72,14 @@ class ObjectPermissionChecker(object):
         :param obj: Django model instance for which permission should be checked
 
         """
-        perm = perm.split('.')[-1]
         if self.user and not self.user.is_active:
             return False
         elif self.user and self.user.is_superuser:
             return True
+        if isinstance(perm, Permission):
+            perm = perm.codename
+        else:
+            perm = perm.split('.')[-1]
         return perm in self.get_perms(obj)
 
     def get_group_filters(self, obj):
diff --git a/guardian/managers.py b/guardian/managers.py
index c16b313..e4c525f 100644
--- a/guardian/managers.py
+++ b/guardian/managers.py
@@ -23,7 +23,7 @@ class BaseObjectPermissionManager(models.Manager):
 
 class UserObjectPermissionManager(BaseObjectPermissionManager):
 
-    def assign_perm(self, perm, user, obj):
+    def assign_perm(self, perm, user, obj, save=True):
         """
         Assigns permission with given ``perm`` for an instance ``obj`` and
         ``user``.
@@ -44,8 +44,10 @@ class UserObjectPermissionManager(BaseObjectPermissionManager):
             kwargs['object_pk'] = obj.pk
         else:
             kwargs['content_object'] = obj
-        obj_perm, created = self.get_or_create(**kwargs)
-        return obj_perm
+        if save:
+            obj_perm, _ = self.get_or_create(**kwargs)
+            return obj_perm
+        return self.model(**kwargs)
 
     def assign(self, perm, user, obj):
         """ Depreciated function name left in for compatibility"""
@@ -77,7 +79,7 @@ class UserObjectPermissionManager(BaseObjectPermissionManager):
 
 class GroupObjectPermissionManager(BaseObjectPermissionManager):
 
-    def assign_perm(self, perm, group, obj):
+    def assign_perm(self, perm, group, obj, save=False):
         """
         Assigns permission with given ``perm`` for an instance ``obj`` and
         ``group``.
diff --git a/guardian/shortcuts.py b/guardian/shortcuts.py
index ec96676..bda856a 100644
--- a/guardian/shortcuts.py
+++ b/guardian/shortcuts.py
@@ -2,6 +2,7 @@
 Convenient shortcuts to manage or check object permissions.
 """
 from __future__ import unicode_literals
+from guardian.core import ObjectPermissionChecker
 
 from django.contrib.auth.models import Group
 from django.contrib.auth.models import Permission
@@ -124,9 +125,18 @@ def bulk_assign_perm(perm, user_or_group, queryset):
         ctype = ContentType.objects.get_for_model(queryset.model)
         perm = Permission.objects.get(content_type=ctype, codename=perm)
 
+    checker = ObjectPermissionChecker(user_or_group)
+    checker.prefetch_perms(queryset)
+
+    manager_kwargs = {('user' if user else 'group'): user or group,
+                      'perm': perm,
+                      'save': False}
+
     assigned_perms = []
     for instance in queryset:
-        assigned_perms.append(model.objects.assign_perm(perm, user or group, instance))
+        if not checker.has_perm(perm, instance):
+            assigned_perms.append(model.objects.assign_perm(obj=instance, **manager_kwargs))
+    model.objects.bulk_create(assigned_perms)
     return assigned_perms
 
 
@@ -195,13 +205,29 @@ def bulk_remove_perm(perm, user_or_group, queryset):
     user, group = get_identity(user_or_group)
     perm = perm.split('.')[-1]
 
+    ctype = None
+
     if user:
         model = get_user_obj_perms_model(queryset.model)
+        object_filter = Q(user=user_or_group)
     if group:
         model = get_group_obj_perms_model(queryset.model)
+        object_filter = object_filter & Q(group=user_or_group)
 
-    for instance in queryset:
-        model.objects.remove_perm(perm, user or group, instance)
+    if not isinstance(perm, Permission):
+        perm = perm.split('.')[-1]
+        ctype = ContentType.objects.get_for_model(queryset.model)
+        perm = Permission.objects.get(content_type=ctype, codename=perm)
+
+    if model.objects.is_generic():
+        ctype = ctype or ContentType.objects.get_for_model(queryset.model)
+        object_filter &= Q(content_type=ctype)
+        object_filter &= Q(object_pk__in=queryset)
+    else:
+        object_filter &= Q(content_object__in=queryset)
+
+    object_filter &= Q(permission=perm)
+    model.objects.filter(object_filter).delete()
 
 
 def get_perms(user_or_group, obj):
```

I feel added code should be moved to managers, so it don't merge ready yet. 

```bulk_assign_perm('view_post', u, Post.objects.all()[:10])``` queries:
```
(0.000) SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE ("django_content_type"."model" = 'post' AND "django_content_type"."app_label" = 'posts'); args=('post', 'posts')
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'view_post' AND "auth_permission"."content_type_id" = 9); args=(u'view_post', 9)
(0.000) SELECT "posts_post"."id" FROM "posts_post" LIMIT 10; args=()
(0.000) SELECT "posts_post"."id", "posts_post"."title", "posts_post"."slug", "posts_post"."content", "posts_post"."created_at" FROM "posts_post" LIMIT 10; args=()
(0.000) SELECT "guardian_userobjectpermission"."id", "guardian_userobjectpermission"."permission_id", "guardian_userobjectpermission"."content_type_id", "guardian_userobjectpermission"."object_pk", "guardian_userobjectpermission"."user_id", "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "guardian_userobjectpermission" INNER JOIN "auth_permission" ON ("guardian_userobjectpermission"."permission_id" = "auth_permission"."id") WHERE ("guardian_userobjectpermission"."object_pk" IN ('1', '2', '3', '4', '5', '6', '7', '8', '9', '10') AND "guardian_userobjectpermission"."user_id" = 32 AND "guardian_userobjectpermission"."content_type_id" = 9); args=(u'1', u'2', u'3', u'4', u'5', u'6', u'7', u'8', u'9', u'10', 32, 9)
(0.000) SELECT "guardian_groupobjectpermission"."id", "guardian_groupobjectpermission"."permission_id", "guardian_groupobjectpermission"."content_type_id", "guardian_groupobjectpermission"."object_pk", "guardian_groupobjectpermission"."group_id", "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "guardian_groupobjectpermission" INNER JOIN "auth_group" ON ("guardian_groupobjectpermission"."group_id" = "auth_group"."id") INNER JOIN "core_customuser_groups" ON ("auth_group"."id" = "core_customuser_groups"."group_id") INNER JOIN "auth_permission" ON ("guardian_groupobjectpermission"."permission_id" = "auth_permission"."id") WHERE ("core_customuser_groups"."customuser_id" = 32 AND "guardian_groupobjectpermission"."content_type_id" = 9 AND "guardian_groupobjectpermission"."object_pk" IN ('1', '2', '3', '4', '5', '6', '7', '8', '9', '10')); args=(32, 9, u'1', u'2', u'3', u'4', u'5', u'6', u'7', u'8', u'9', u'10')
(0.000) BEGIN; args=None
(0.001) INSERT INTO "guardian_userobjectpermission" ("permission_id", "content_type_id", "object_pk", "user_id") SELECT 28, 9, '1', 32 UNION ALL SELECT 28, 9, '2', 32 UNION ALL SELECT 28, 9, '3', 32 UNION ALL SELECT 28, 9, '4', 32 UNION ALL SELECT 28, 9, '5', 32 UNION ALL SELECT 28, 9, '6', 32 UNION ALL SELECT 28, 9, '7', 32 UNION ALL SELECT 28, 9, '8', 32 UNION ALL SELECT 28, 9, '9', 32 UNION ALL SELECT 28, 9, '10', 32; args=(28, 9, u'1', 32, 28, 9, u'2', 32, 28, 9, u'3', 32, 28, 9, u'4', 32, 28, 9, u'5', 32, 28, 9, u'6', 32, 28, 9, u'7', 32, 28, 9, u'8', 32, 28, 9, u'9', 32, 28, 9, u'10', 32)
```
```bulk_remove_perm('view_post', u, Post.objects.all()[:10])``` queries:
```
(0.000) SELECT "auth_permission"."id", "auth_permission"."name", "auth_permission"."content_type_id", "auth_permission"."codename" FROM "auth_permission" WHERE ("auth_permission"."codename" = 'view_post' AND "auth_permission"."content_type_id" = 9); args=(u'view_post', 9)
(0.000) BEGIN; args=None
(0.000) DELETE FROM "guardian_userobjectpermission" WHERE ("guardian_userobjectpermission"."user_id" = 32 AND "guardian_userobjectpermission"."content_type_id" = 9 AND "guardian_userobjectpermission"."object_pk" IN (SELECT "posts_post"."id" FROM "posts_post" LIMIT 10) AND "guardian_userobjectpermission"."permission_id" = 28); args=(32, 9, 28)
```

We have to consider #366 "regression" too.

**Comment by ad-m** [comment]
I continue work https://github.com/ad-m/django-guardian/tree/pr-432 . I do pull requests when ready. I have to add tests for updated API for pass ```Permission``` instance as ```perm``` param.

**Comment by brianmay** [comment]
@ad-m So are you saying I shouldn't merge this just yet? If so, let me know when you are ready.

**Comment by brianmay** [comment]
Merged in #436. Closing.

**Pull request closed without merging**